File: rust/11.container-with-most-water.rs
/*
 * @lc app=leetcode id=11 lang=rust
 *
 * [11] Container With Most Water
 */

// @lc code=start
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut result = 0;
        let mut iter = height.iter().enumerate();
        let (mut left, mut right) = (iter.next(), iter.next_back());
        while let (Some((l, h1)), Some((r, h2))) = (left, right) {
            let w = (r - l) as i32;
            result = result.max(h1.min(h2) * w);
            if h1 < h2 {
                left = iter.next();
            } else {
                right = iter.next_back();
            }
        }
        return result;
    }
}
// @lc code=end

File: javascript/11.container-with-most-water.js
/*
 * @lc app=leetcode id=11 lang=javascript
 *
 * [11] Container With Most Water
 */

// @lc code=start
/**
 * @param {number[]} height
 * @return {number}
 */
var maxArea = function (height) {
  let mostContainer = 0;
  let l = 0,
    r = height.length - 1;
  while (l < r) {
    const w = r - l;
    const h = height[l] < height[r] ? height[l++] : height[r--];
    mostContainer = Math.max(w * h, mostContainer);
  }
  return mostContainer;
};
// @lc code=end
